add try/catch to fromData, move completed stalled jobs to completed

howardchung opened the pull request:
This change should stop processStalledJobs from getting an error, which causes it to get stuck due to no catch.

howardchung commented:
Any problems with this pull request?

howardchung commented:
@manast ?

manast commented:
the problem is that there are too many changes not related to the title of the pull request... if you could break it down so that we know why there are there it would be much easier to merge.

howardchung commented:
rebased and squashed into one commit.  There are only two changes:

* try/catch the fromData.  If the JSON parse fails, remove the job.
* Move stalled completed jobs to completed.

howardchung commented:
synced to master.  Shouldn't this fix #273 ?

chuym commented on lib/job.js:
```diff
@@ -393,8 +393,19 @@ Job.prototype._saveAttempt = function(){
+  }
+  catch (err) {
+    //Job unable to be parsed. Remove it.
+    var job = new Job(queue, JSON.parse(data.data), JSON.parse(data.opts));
```
The above throws because `JSON.parse` attempts to parse undefined. Wouldn't this throw the error again?

howardchung replied:
Yeah. It should be {},{}

chuym commented on lib/queue.js:
```diff
@@ -507,6 +507,9 @@ Queue.prototype.processStalledJob = function(job){
             return _this.processJob(job, true);
           }
+          else {
+            return job.moveToCompleted();
```
If this PR is fixing #273 then this line should be on a separate PR.

manast commented:
I am in the doubt if we should move corrupted jobs to the failed set or sanitize them, I am leaning more towards the failed jobs solution...
